### PR TITLE
fix(nuget): start with empty nuget.config for updating lockfile

### LINKED ISSUE #7769: Error updating NuGet lock file when credentials are defined for a feed
**What Renovate type, platform and version are you using?**

Self-hosted v23.84.6 running in Bitbucket Pipelines.

**Describe the bug**

Renovate always fails to update lock file when a private feed requiring credentials is defined in nuget.config. The problem is pretty obvious from the logs - the command `dotnet nuget update source` is run with empty string as the value of username and password, but this command is not supported and throws an error.

**Relevant debug logs**

_Note: I redacted some of the data._

```
DEBUG: dotnet command (repository=**redacted**, branch=**redacted**)
       "cmd": [
         "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text",
         "dotnet restore **redacted** --force-evaluate",
         "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text"
       ]
DEBUG: Executing command (repository=**redacted**, branch=**redacted**)
       "command": "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text"
DEBUG: exec completed (repository=**redacted** , branch=**redacted** )
       "cmd": "dotnet nuget update source **redacted** --username **redacted** --password **redacted** --store-password-in-clear-text",
       "durationMs": 442,
       "stdout": "Package source \"**redacted**\" was successfully updated.\n",
       "stderr": ""
DEBUG: Executing command (repository=**redacted** , branch=**redacted** )
       "command": "dotnet restore **redacted**  --force-evaluate"
DEBUG: exec completed (repository=**redacted** , branch=**redacted**)
       "cmd": "dotnet restore **redacted** --force-evaluate",
       "durationMs": 2422,
       "stdout": "  Determining projects to restore...\n  Restored **redacted** (in 947 ms).\n",
       "stderr": ""
DEBUG: Executing command (repository=**redacted**, branch=**redacted**)
       "command": "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text"
DEBUG: Failed to generate lock file (repository=**redacted** , branch=**redacted** )
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text",
         "stdout": "",
         "stderr": "System.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n",
         "message": "Command failed: dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text\nSystem.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n",
         "stack": "Error: Command failed: dotnet nuget update source **redacted** --username '' --password '' --store-password-in-clear-text\nSystem.ArgumentException: Value cannot be null or whitespace. (Parameter 'token')\n   at Microsoft.DotNet.Cli.CommandLine.OptionError..ctor(String message, String token, AppliedOption option) in /_/src/source/CommandLine/OptionError.cs:line 21\n   at Microsoft.DotNet.Cli.CommandLine.Parser.UnrecognizedArg(String arg) in /_/src/source/CommandLine/Parser.cs:line 170\n   at System.Linq.Enumerable.SelectListIterator`2.MoveNext()\n   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)\n   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(IReadOnlyCollection`1 rawArgs, Boolean isProgressive) in /_/src/source/CommandLine/Parser.cs:line 111\n   at Microsoft.DotNet.Cli.CommandLine.Parser.Parse(String[] args) in /_/src/source/CommandLine/Parser.cs:line 31\n   at Microsoft.DotNet.Cli.ParserExtensions.ParseFrom(Parser parser, String context, String[] args)\n   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)\n   at Microsoft.DotNet.Cli.Program.Main(String[] args)\n\n    at ChildProcess.exithandler (child_process.js:308:12)\n    at ChildProcess.emit (events.js:314:20)\n    at ChildProcess.EventEmitter.emit (domain.js:483:12)\n    at maybeClose (internal/child_process.js:1021:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)"
       }

```

**To Reproduce**

1. Create a .NET Core project and reference an outdated package.
2. Add a `nuget.config` file defining a feed that requires username and password:

   ```xml
   <configuration>
     <packageSources>
       <clear />
       <add key="nuget.org"
          value="https://api.nuget.org/v3/index.json" />
       <add key="PrivateFeed"
          value="my-private-feed-url" />
     </packageSources>
   </configuration>
   ```

3. Set credentials in Renovate `hostRules` config:

   ```json
   {
     "hostRules": [
       {
         "timeout": 60000,
         "baseUrl": "my-private-feed-url",
         "hostType": "nuget",
         "username": "the-username",
         "password": "the-password"
       }
     ]
   }
   ```

I could not provide an example repo because I can't share credentials to a private feed...

**Additional context**

NuGet lock file support was added in #7375.

From a quick search in code, it seems the command with empty user and pass is run by [lib/manager/nuget/artifacts.ts:37](https://github.com/renovatebot/renovate/blob/master/lib/manager/nuget/artifacts.ts#L37).

### PR DESCRIPTION
**Pull request by fgreinacher**
The previously used command, `dotnet new nugetconfig` creates a nuget.config with some defaults that we are not interested in while creating the temporary file for authentication.

Fixes #7769

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added unit tests, or
- [ ] Unit tests + ran on a real repository

### CODE REVIEW

File: lib/manager/nuget/util.ts [hunk 1]
```diff
@@ -25,7 +25,16 @@ export function getRandomString(): string {
...
+export function getDefaultRegistries(): Registry[] {
+  return datasourceNuget.defaultRegistryUrls.map(
+    (registryUrl) =>
+      ({
+        url: registryUrl,
+      } as Registry)
+  );
+}
+
+export async function getConfiguredRegistries(
```
Comment (fgreinacher) [inline lib/manager/nuget/util.ts:37]: I renamed `determineRegistries` to `getConfiguredRegistries` to make it clearer that this just returns something if there is a NuGet config. It's a bit weird, as actually I would have expected this to return the default registries if there is no config file, but I think we did it for back-compatibility back then.

File: lib/manager/nuget/util.ts [hunk 2]
```diff
@@ -56,12 +65,7 @@ export async function determineRegistries(
     return undefined;
   }
 
-  const registries = datasourceNuget.defaultRegistryUrls.map(
```
Comment (fgreinacher) [inline lib/manager/nuget/util.ts:59]: Moved to separate function so that I could re-use it when filling the temporary config file.

**Review by viceice — COMMENTED** [review]
Needs a real run to be sure it works as expected

**Comment by fgreinacher** [comment]
I set up [a sample repo](https://github.com/fgreinacher/tmp-renovate-private-nuget-feed) and ran Renovate with:

```
export LOG_LEVEL=debug
export RENOVATE_HOST_RULES="[ { \"hostName\": \"example.org\", \"username\": \"example-user\",  \"password\": \"example-password\" } ]"
yarn start fgreinacher/tmp-renovate-private-nuget-feed
```

Log output does not show any issues:

```
DEBUG: nuget.updateArtifacts(test.csproj) (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
DEBUG: found NuGet.config (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "nuGetConfigPath": "/var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/repos/github/fgreinacher/tmp-renovate-private-nuget-feed/nuget.config"
DEBUG: clearing registry URLs (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
DEBUG: adding registry URL (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "registryUrl": "https://api.nuget.org/v3/index.json"
DEBUG: adding registry URL (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "registryUrl": "https://example.org"
DEBUG: dotnet command (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "cmd": [
         "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name nuget",
         "dotnet nuget add source https://example.org --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name example --username example-user --password **redacted** --store-password-in-clear-text",
         "dotnet restore test.csproj --force-evaluate --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config"
       ]
DEBUG: Executing command (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "command": "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name nuget"
DEBUG: exec completed (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "cmd": "dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name nuget",
       "durationMs": 460,
       "stdout": "Package source with Name: nuget added successfully.\n",
       "stderr": ""
DEBUG: Executing command (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "command": "dotnet nuget add source https://example.org --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name example --username example-user --password **redacted** --store-password-in-clear-text"
DEBUG: exec completed (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "cmd": "dotnet nuget add source https://example.org --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config --name example --username example-user --password **redacted** --store-password-in-clear-text",
       "durationMs": 453,
       "stdout": "Package source with Name: example added successfully.\n",
       "stderr": ""
DEBUG: Executing command (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "command": "dotnet restore test.csproj --force-evaluate --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config"
DEBUG: exec completed (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
       "cmd": "dotnet restore test.csproj --force-evaluate --configfile /var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/cache/others/nuget/db9e2cacddea8f02/nuget.config",
       "durationMs": 1654,
       "stdout": "  Determining projects to restore...\n  Restored /private/var/folders/n2/p5n3nc_56cl3lxjxkkc1_dsr0000gn/T/renovate/repos/github/fgreinacher/tmp-renovate-private-nuget-feed/test.csproj (in 201 ms).\n",
       "stderr": ""
DEBUG: Returning updated lock file (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)
DEBUG: Updated 1 package files (repository=fgreinacher/tmp-renovate-private-nuget-feed, branch=renovate/system.io.abstractions-13.x)

```

And the PR is also looking perfectly fine: https://github.com/fgreinacher/tmp-renovate-private-nuget-feed/pull/4

**Comment by fgreinacher** [comment]
(BTW: Do you think it would make sense to introduce some kind of high-level testing against real repos and registries?)

**Comment by viceice** [comment]
> (BTW: Do you think it would make sense to introduce some kind of high-level testing against real repos and registries?)

I would like to avoid this, as this will slow down ci testing even more 😕

**Review by viceice — APPROVED** [review]

**Comment by fgreinacher** [comment]
macOS job ran into timeout

**Comment by fgreinacher** [comment]
Friendly ping @rarkins 😀

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
https://www.githubstatus.com/incidents/sxthxvdlljk4 😕

**Review by rarkins — APPROVED** [review]

**Comment by lazyboy1** [comment]
It works! 🎉 Thanks @fgreinacher

**Comment by lazyboy1** [comment]
It works! 🎉 Thanks @fgreinacher

**Pull request merged**